File: backend/internal/service/tests_04.go
package service


import (
	"context"
	"time"

	"github.com/biruk-ak/clinical-lis/backend/internal/models"
	"github.com/biruk-ak/clinical-lis/backend/internal/repository"
	"github.com/google/uuid"
)

// LabTestRecord04Service encapsulates business rules for LabTestRecord04.
type LabTestRecord04Service struct {
	repo *repository.LabTestRecord04Repository
}

// NewLabTestRecord04Service wires the service.
func NewLabTestRecord04Service(repo *repository.LabTestRecord04Repository) *LabTestRecord04Service {
	return &LabTestRecord04Service{repo: repo}
}

// Create validates and persists a new LabTestRecord04.
func (s *LabTestRecord04Service) Create(ctx context.Context, actor, facilityID, code, display, status string, priority int, notes string) (*models.LabTestRecord04, error) {
	now := time.Now().UTC()
	rec := &models.LabTestRecord04{
		ID:           uuid.NewString(),
		ExternalCode: code,
		DisplayName:  display,
		Status:       status,
		Priority:     priority,
		FacilityID:   facilityID,
		CreatedBy:    actor,
		UpdatedBy:    actor,
		Notes:        notes,
		MetadataJSON: "{}",
		Version:      1,
		IsActive:     true,
		CreatedAt:    now,
		UpdatedAt:    now,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	if err := s.repo.Create(ctx, rec); err != nil {
		return nil, err
	}
	return rec, nil
}

// Get returns a LabTestRecord04 by id.
func (s *LabTestRecord04Service) Get(ctx context.Context, id string) (*models.LabTestRecord04, error) {
	return s.repo.GetByID(ctx, id)
}

// List returns facility-scoped LabTestRecord04 pages.
func (s *LabTestRecord04Service) List(ctx context.Context, facilityID string, limit, offset int) ([]models.LabTestRecord04, error) {
	if limit <= 0 || limit > 200 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}
	return s.repo.List(ctx, facilityID, limit, offset)
}

// Update mutates an existing LabTestRecord04.
func (s *LabTestRecord04Service) Update(ctx context.Context, actor string, rec *models.LabTestRecord04) error {
	if err := rec.Validate(); err != nil {
		return err
	}
	rec.Touch(actor, time.Now().UTC())
	return s.repo.Update(ctx, rec)
}

// Archive soft-deletes a LabTestRecord04.
func (s *LabTestRecord04Service) Archive(ctx context.Context, id, actor string) error {
	return s.repo.Delete(ctx, id, actor)
}

// Search finds LabTestRecord04 by external code prefix.
func (s *LabTestRecord04Service) Search(ctx context.Context, facilityID, prefix string) ([]models.LabTestRecord04, error) {
	return s.repo.SearchByCode(ctx, facilityID, prefix, 100)
}

// Stats returns active counts for dashboards.
func (s *LabTestRecord04Service) Stats(ctx context.Context, facilityID string) (int64, error) {
	return s.repo.CountActive(ctx, facilityID)
}

File: frontend/src/types/tests04.ts
/** Domain types for tests / LabTestRecord04 */
export type LabTestRecord04Status = 'draft' | 'active' | 'pending' | 'completed' | 'cancelled' | 'archived';

export interface LabTestRecord04 {
  id: string;
  externalCode: string;
  displayName: string;
  status: LabTestRecord04Status;
  priority: number;
  facilityId: string;
  createdBy: string;
  updatedBy: string;
  notes: string;
  metadataJson: string;
  version: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string | null;
}

export interface LabTestRecord04CreateRequest {
  externalCode: string;
  displayName: string;
  status: LabTestRecord04Status;
  priority: number;
  facilityId: string;
  notes?: string;
}

export interface LabTestRecord04ListResponse {
  items: LabTestRecord04[];
  total?: number;
}

export interface LabTestRecord04StatsResponse {
  activeCount: number;
}

export function isLabTestRecord04Active(rec: LabTestRecord04): boolean {
  return rec.isActive && rec.status !== 'archived' && rec.status !== 'cancelled';
}

export function formatLabTestRecord04Label(rec: LabTestRecord04): string {
  return `${rec.externalCode} — ${rec.displayName}`;
}

export function compareLabTestRecord04ByPriority(a: LabTestRecord04, b: LabTestRecord04): number {
  if (a.priority !== b.priority) return b.priority - a.priority;
  return a.displayName.localeCompare(b.displayName);
}
